content/blog/2017/2017-02-14-declarative-notifications.adoc

```diff
@@ -0,0 +1,436 @@
+I've set up a Multibranch Pipeline and pointed it at my repository
+the same as did it previous post.
+Also the same as before, I've set this Pipeline's Git configuration to
+automatically "Clean after checkout".
+
+In addition, I still have my notification targets setup from the
+"link:/blog/2016/07/18/pipline-notifications/[Sending Notifications in Pipeline]" blog post.
+Take a look there for help setting up the
+plugin:slack[Slack],
+plugin:hipchat[HipChat],
+and plugin:email-ext[Email-ext]
+plugins.
+
+
+Here's the final form of the Declarative Pipeline from the previous post:
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent {
+    // Use docker container
+    docker {
+      image 'ruby:2.3'
+    }
+  }
+  options {
+    // Only keep the 10 most recent builds
+    buildDiscarder(logRotator(numToKeepStr:'10'))
+  }
+  stages {
+    stage ('Install') {
+      steps {
+        // install required gems
+        sh 'bundle install'
+      }
+    }
+    stage ('Build') {
+      steps {
+        // build
+        sh 'bundle exec rake build'
+      }
+
+      post {
+        success {
+          // Archive the built artifacts
+          archive includes: 'pkg/*.gem'
+        }
+      }
+    }
+    stage ('Test') {
+      steps {
+        // run tests with coverage
+        sh 'bundle exec rake spec'
+      }
+
+      post {
+        success {
+          // publish html
+          publishHTML target: [
+              allowMissing: false,
+              alwaysLinkToLastBuild: false,
+              keepAll: true,
+              reportDir: 'coverage',
+              reportFiles: 'index.html',
+              reportName: 'RCov Report'
+            ]
+        }
+      }
+    }
+  }
+  post {
+    always {
+      echo "Send notifications for result: ${currentBuild.result}"
+    }
+  }
+}
+// Scripted //
+----
+
+
+== Preparing for Notifications
+
+The Pipeline above has a final `always` block that prints a message to the status log,
+but doesn't print anything at the start of the run.
+The final form of the Pipeline in the original blog post called the same method twice,
+once at the start and once at the end of the run.
+Let's start by doing that here as well.
+
+[source, groovy]
+----
+pipeline {
+  /* ... unchanged ... */
+  stages {
+    stage ('Start') {
+      steps {
+        // send build started notifications
+        sendNotifications 'STARTED'
+      }
+    }
+    /* ... unchanged ... */
+  }
+  post {
+    always {
+      sendNotifications currentBuild.result
+    }
+  }
+}
+
+def sendNotifications(String buildStatus = 'STARTED') {
+  echo "Send notifications for run: ${buildStatus}"
+}
+----
+
+.Declarative is still Jenkins Pipeline
+****
+Jenkins Declarative Pipeline is still *Jenkins*, still *Pipeline*, and still *Groovy*.
+Declarative Pipeline - everything inside
+the `pipeline {}` block - is a domain-specific subset of Groovy, but outside that
```

Comment:
"domain specific subset of Groovy" I don't think that's a good statement to make. It's using a pre-defined structure that must be valid Groovy syntax, but to say that it's a subset of Groovy I think is disingenuous and is going to lead to confusion.


What you could say instead is that, underneath the covers, Jenkins Pipeline uses the Groovy engine to parse the `Jenkinsfile` and while everything inside of `pipeline {}` must follow a strict structure and syntax, outside of the block you can actually insert Groovy code. **ASSUMING** @abayer is comfortable with supporting this "feature" in perpetuity :smile: 


Otherwise, highlighting this might hurt Pipeline implementation further down the line.


There's also the potential issues with the Pipeline Visual Editor, I'm not sure how it treats code outside of the `pipeline {}` block. If it doesn't treat it well, then we might have to retool, this blog post.